### PR TITLE
Fix modeling rule type comparison

### PR DESCRIPTION
**Pull request by kobymeir**
## Description
Fix modeling rule type comparison

### CODE REVIEW

File: demisto_sdk/commands/lint/helpers.py [hunk 1]
```diff
@@ -310,7 +310,10 @@ def add_api_modules(
             copied_api_module_path = pack_path / f"{module_name}.py"
             if content_repo:  # if working in a repo
                 module_path = content_repo / api_module_path
-                shutil.copy(src=module_path, dst=copied_api_module_path)
+                try:
+                    os.symlink(src=module_path, dst=copied_api_module_path)
+                except NotImplementedError:
+                    shutil.copy(src=module_path, dst=copied_api_module_path)
```
Comment (kobymeir) [inline demisto_sdk/commands/lint/helpers.py:316]:
```suggestion
                shutil.copy(src=module_path, dst=copied_api_module_path)
```

Comment (kobymeir) [inline demisto_sdk/commands/lint/helpers.py:316]:
```suggestion
                shutil.copy(src=module_path, dst=copied_api_module_path)	
```

File: demisto_sdk/commands/lint/helpers.py [hunk 2]
```diff
@@ -310,7 +310,7 @@ def add_api_modules(
             copied_api_module_path = pack_path / f"{module_name}.py"
             if content_repo:  # if working in a repo
                 module_path = content_repo / api_module_path
-                shutil.copy(src=module_path, dst=copied_api_module_path)
+                shutil.copy(src=module_path, dst=copied_api_module_path)
```
Comment (kobymeir) [inline demisto_sdk/commands/lint/helpers.py:313]:
```suggestion
                shutil.copy(src=module_path, dst=copied_api_module_path)
```

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 3]
```diff
@@ -318,7 +322,7 @@ def verify_results_against_test_data(
                 logger.debug(f"[cyan]{out}[/cyan]", extra={"markup": True})
                 result_test_case_system_out.append(out)
 
-                if type(result_val) == type(val) and result_val == val:
+                if result_val == val:
```
Comment (kobymeir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:325]: the actual bug. comparing the types. there is an issue with int/float comparison of value returned from XSIAM. all numeric values are returned as float type.

Comment (tkatzir) [reply demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:325]: Maybe change the original check to something like:
introduce the following method (pseudocode,) and use it.
```
def get_type_family(object) -> Enum(STRING, NUMBER, DATE, BOOLEAN):
    if type(object) == StringType:
        return STRING
    if type(object) in [intType, floatType]:
        return NUMBER
    # same for DATE and BOOLEAN
```

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 4]
```diff
@@ -750,8 +754,8 @@ def verify_pack_exists_on_tenant(
                 "[red]Pack does not exist on the tenant. Please install or upload the pack and try again[/red]",
                 extra={"markup": True},
             )
-            typer.echo(
-                f"demisto-sdk upload -z -x -i {containing_pack.path}\ndemisto-sdk modeling-rules test {mr.path.parent}"
+            logger.info(
+                f"\ndemisto-sdk upload -z -x -i {containing_pack.path}\ndemisto-sdk modeling-rules test {mr.path.parent}"
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:758]: Why `\n` at the start?

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 5]
```diff
@@ -897,7 +900,7 @@ def validate_modeling_rule(
                         err, schema_test_case, modeling_rule_test_suite
                     )
         else:
-            err = f"Schema file does not exist in path {modeling_rule.schema_path}"
+            err = f"Schema file does not exist in path {get_relative_path_to_content(modeling_rule.schema_path)}"
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:903]: If it's not there, will `get_relative_path_to_content` work?

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 6]
```diff
@@ -1044,17 +1047,15 @@ def validate_modeling_rule(
                 )
                 logger.error(
                     f"[red]Please create a test data file for "
-                    f"{get_relative_path_to_content(modeling_rule_directory)} and then rerun[/red]",
+                    f"{get_relative_path_to_content(modeling_rule_directory)} and then rerun\n{executed_command}[/red]",
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:1050]: If it's not there, will `get_relative_path_to_content` work?

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 7]
```diff
@@ -1044,17 +1047,15 @@ def validate_modeling_rule(
...
                     extra={"markup": True},
                 )
-                typer.echo(executed_command)
         else:
             logger.error(
                 f"[red]Please create a test data file for "
-                f"{get_relative_path_to_content(modeling_rule_directory)} and then rerun[/red]",
+                f"{get_relative_path_to_content(modeling_rule_directory)} and then rerun\n{executed_command}[/red]",
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:1056]: Same comment.
Also: Can the messages be unique?

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 8]
```diff
@@ -1279,16 +1280,26 @@ def test_modeling_rule(
     handle_deprecated_args(ctx.args)
 
     logger.info(
-        f"[cyan]Modeling Rules directories to test: {inputs}[/cyan]",
+        "[cyan]Test Modeling Rules directories to test:[/cyan]",
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:1283]: ??? What are the directories to test?
Did you delete `{inputs}` by mistake?

**Review by tkatzir — COMMENTED** [review]
Good! Not approving yet, some major comments.

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py (hunk 5 above)
Comment (kobymeir) [reply demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:903]: yes, tested the function now. good thought!

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py (hunk 6 above)
Comment (kobymeir) [reply demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:1050]: yes, answered in the other comment

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py (hunk 7 above)
Comment (kobymeir) [reply demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:1056]: will fix

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py (hunk 8 above)
Comment (kobymeir) [reply demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:1283]: i'm printing them one per line, in the next lines. make it more readable in the build logs

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py (hunk 3 above)
Comment (kobymeir) [reply demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:325]: fixed.

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 9]
```diff
@@ -311,44 +339,59 @@ def verify_results_against_test_data(
             )
         table_result = create_table(expected_values, result)
         logger.info(f"\n{table_result}")
-        for key, val in expected_values.items():
-            if val:
-                result_val = result.get(key)
-                out = f"Checking for key {key}:\n - expected: {val}\n - received: {result_val}"
+        for expected_key, expected_value in expected_values.items():
+            if expected_value:
+                received_value_before = result.get(expected_key)
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:344]: Rename:
* received_value_before -> received_value
* received_value (in the tuple result of the call to transform_received_value_by_expected_type) -> received_value_sanitized
* received_value_type -> received_value_sanitized_type

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 10]
```diff
@@ -311,44 +339,59 @@ def verify_results_against_test_data(
...
+                (
+                    received_value_type,
+                    received_value,
+                ) = transform_received_value_by_expected_type(
+                    received_value_before, expected_value_type
+                )
+                out = (
+                    f"Checking for key {expected_key} - "
+                    f"expected:{expected_value} expected type:{expected_value_type} "
+                    f"received:{received_value} received type:{received_value_type} "
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:355]:
```suggestion
                    f"expected value:{expected_value} expected type:{expected_value_type} "
                    f"received value:{received_value} received type:{received_value_type} "
```

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 11]
```diff
@@ -311,44 +339,59 @@ def verify_results_against_test_data(
...
+                            f"received: {received_value} Types match:{received_value_type}"
                         )
                         logger.error(
-                            f'[red][bold]{key}[/bold] --- "{result_val}" != "{val}" '
-                            f"Types match:{get_type_pretty_name(result_val)}[/red]",
+                            f'[red][bold]{expected_key}[/bold] --- "{received_value}" != "{expected_value}" '
+                            f"Types match:{received_value_type}[/red]",
                             extra={"markup": True},
                         )
                     else:
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:379]: Document this `else`. Something like:
```suggestion
                    else:
                        # Types don't match, so values are not matching either,
                        # so it means that both do not match.
```

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 12]
```diff
@@ -311,44 +339,59 @@ def verify_results_against_test_data(
...
+                            f'[red][bold]{expected_key}[/bold][red] --- "{received_value}" != "{expected_value}"\n'
+                            f' [bold]{expected_key}[/bold][red] --- Received value type: "{received_value_type}" '
+                            f'!= Expected value type: "{expected_value_type}"[/red]',
                             extra={"markup": True},
                         )
                     result_test_case_system_err.append(err)
                     result_test_case_results.append(Failure(err))
             else:
-                err = f"No mapping for this {key} - skipping checking match"
+                err = f"No mapping for this {expected_key} - skipping checking match"
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:394]:
```suggestion
                err = f"No mapping for key {expected_key} - skipping checking match"
```

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 13]
```diff
@@ -1302,21 +1352,44 @@ def test_modeling_rule(
...
+        if success:
+            logger.info(
+                f"[green]Test Modeling rule {get_relative_path_to_content(modeling_rule_directory)} passed[/green]",
+                extra={"markup": True},
+            )
+        else:
             errors = True
             logger.error(
-                f"[red]Error testing modeling rule {get_relative_path_to_content(modeling_rule_directory)}[/red]",
+                f"[red]Test Modeling rule {get_relative_path_to_content(modeling_rule_directory)} failed[/red]",
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:1363]: Either Modeling Rule, or modeling rule. I like the capitalization.
```suggestion
                f"[red]Test Modeling Rule {get_relative_path_to_content(modeling_rule_directory)} failed[/red]",
```

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py [hunk 14]
```diff
@@ -1302,21 +1352,44 @@ def test_modeling_rule(
...
                 extra={"markup": True},
             )
         if modeling_rule_test_suite:
             modeling_rule_test_suite.add_property("start_time", start_time)
             xml.add_testsuite(modeling_rule_test_suite)
 
     if output_junit_file:
+        logger.info(
+            f"[cyan]Writing JUnit XML to {output_junit_file}[/cyan]",
```
Comment (tkatzir) [inline demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:1372]: JUnit is Java, but we won't change it now. It is what it is...

**Review by tkatzir — APPROVED** [review]
Good!
Approved, with minor comments.

File: demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py (hunk 14 above)
Comment (kobymeir) [reply demisto_sdk/commands/test_content/test_modeling_rule/test_modeling_rule.py:1372]: https://www.ibm.com/docs/en/developer-for-zos/14.1?topic=formats-junit-xml-format
junit is a file type format. not just for java.

**Pull request merged**